### PR TITLE
WIP - Kernel API design

### PR DESCRIPTION
**Pull request by mzient**
The proposed kernel API is as follows:
The kernel developer provides a class with two static methods:
```
void Run(KernelContext &context, [Outputs], [Inputs], [ExtraArgs])
KernelRequirements GetRequirements(KernelContext &context, [Inputs], [ExtraArgs])
```
Outputs, Inputs and ExtraArgs are just ordinary lists of function parameters.
Outputs are TensorLists with non-const element type
Inputs are TensorLists with const element type.
Everything else is an considered an extra argument.

`GetRequirements` provides output shapes, as calculated based on input (shape) and extra arguments.
`Run` is meant to perform the actual work.
`KernelContext` will contain a cuda stream and possibly some information for sub-batch procesing, for cases when we want to execute GetRequirements for the whole batch, but execute in minibatches. This is still a matter to discuss.

Helper methods `kernel::Run` and `kernel::GetRequirements` operate on tuples of inputs, outputs and extra arguments. Also, these tuples can be extracted using type traits:
`kernel_inputs<KernelType>` - searches the Run function signature and extracts all inputs
`kernel_outputs<KernelType>` - same, but for outputs
`kernel_args<KernelType>` - likewise, for extra args

Instantiating `check_kernel<KernelType>` for given kernel will trigger a cascade of compile-time checks that verify that the type satisifies conditions to be considered a kernel - i.e. checks for existence of unique, static methods Run and GetRequirements and verify that parameters and their order is preserved. These checks provide meaningful diagnostics using `static_assert`s.

kernel::Run has an overload taking a vector of inputs and a vector of outputs and is meant to implement the functionality of processing multiple input sets. The default implementations simply executes the kernel for each respective pair of inputs and outputs.

Finally (?)
`TYPE_SWITCH` and `VALUE_SWITCH` mechanisms provide dynamic-to-static type and value routing in a way similar to `DALI_TYPE_SWITCH`, but have capability to handle different set of types/values at each invocation. The `case` code for these switches provides either a typedef (TYPE_SWITCH) or a const auto (for VALUE_SWITCH), which can be used e.g. for template instantiation. Actual type-to-id mapping trait is configurable, as is handling of default case (to decouple the library code from DALI_ENFORCE, which is quite involved).

### CODE REVIEW

File: dali/kernels/tensor_view.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "shape.h"
+
+namespace dali {
+
+template <typename StorageBackend, typename DataType, size_t dim>
```
Comment (Kh4L) [inline dali/kernels/tensor_view.h]: What is the StorageBackend? GPU/CPU or the Allocator class?

File: dali/kernels/tensor_view.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_TENSOR_VIEW_
+#define DALI_KERNELS_TENSOR_VIEW_
+
+
+#include "shape.h"
+#include "backend_tags.h"
```
Comment (Kh4L) [inline dali/kernels/tensor_view.h]: What is this header?

Comment (mzient) [reply dali/kernels/tensor_view.h]: Missing; added.

File: dali/kernels/tensor_view.h (hunk 1 above)
Comment (mzient) [reply dali/kernels/tensor_view.h]: Outdated.

**Comment by szalpal** [comment]
If I understand correcty:

> A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function. 
> *https://en.cppreference.com/w/cpp/language/inline*

some `inline`s are redundant here

File: dali/kernels/shape.h [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SHAPE_H_
+#define DALI_KERNELS_SHAPE_H_
+
+#include <dali/common.h>
+#include <vector>
+#include <array>
+
+namespace dali {
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/shape.h]: If so, I'd certainly use a nested namespace dali::kernels. However, currently we have just one namespace for everything and putting kernels somewhere else would stand out.

File: dali/CMakeLists.txt [hunk 4]
```diff
@@ -44,7 +45,7 @@ endif()
 set(CUDA_LINK_LIBRARIES_KEYWORD PRIVATE)
 if(${CMAKE_VERSION} VERSION_LESS "3.9.0")
   # In cmake 3.8.0 or lower, CUDA_LINK_LIBRARIES_KEYWORD is not available
-  # So this is a workaround to append PRIVATE in front of it 
+  # So this is a workaround to append PRIVATE in front of it
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/CMakeLists.txt]: Ubuntu 16.04 got <  3.9.0 by default.

**Comment by JanuszL** [comment]
> If I understand correcty:
> 
> > A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function.
> > _https://en.cppreference.com/w/cpp/language/inline_
> 
> some `inline`s are redundant here

On the other hand if someone decide moving them into separate *.cc file then he know that he should keep them inlined.

File: dali/kernels/kernel.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+#include "tensor_view.h"
+#include "context.h"
+
+namespace dali {
+
+template <typename InputType, typename OutputType>
+struct Kernel {
+
+  struct Inputs {
+    TensorListView<StorageBackend::GPU, const InputType, 3> images;
```
Comment (JanuszL) [inline dali/kernels/kernel.h]: I would call this just data, as we can process bboxes or other data formats as well.

File: dali/kernels/shape.h [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SHAPE_H_
+#define DALI_KERNELS_SHAPE_H_
+
+#include <dali/common.h>
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/shape.h]: 1. Operating on signed values is generally more natural. Handling boundary cases (clamping, wrapping, etc) is much easier with signed values.
2. We must rethink our layout of common files and utilities. Maybe we should have them grouped in dali/common or dali/util and make that available to both the pipeline and kernel library.

File: dali/kernels/tensor_view.h [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  static const size_t static_dim = dim_;
+  size_t dim() const { return ShapeDim(shape); }
+
+  template <typename Offset>
+  DataType *at(const Offset &pos) const {
+    return data + CalcOfffset(shape, pos);
+  }
+
+  template <typename... Indices>
+  DataType *at(Index idx0, Indices&& ...idx) const {
```
Comment (klecki) [inline dali/kernels/tensor_view.h]: A bit out of scope, but do we want to constrain number of idx passed here to `dim_`? At least in some cases.

File: dali/kernels/tuple_helpers.h [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+struct tuple_cat_helper<t0, tuples...>
+ : tuple_cat_helper<t0, typename tuple_cat_helper<tuples...>::type> {};
+
+template <typename... T1, typename... T2>
+struct tuple_cat_helper<std::tuple<T1...>, std::tuple<T2...>> {
+  using type = std::tuple<T1..., T2...>;
+};
+
+template <typename... tuples>
+using tuple_cat_t = typename tuple_cat_helper<tuples...>::type;
```
Comment (klecki) [inline dali/kernels/tuple_helpers.h]: delctype(std::tuple_cat(std::declval<tuples>{}...)) ?

File: dali/kernels/context.h [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+template <ComputeBackend>
+struct Context {};
+
+template <>
+struct Context<ComputeBackend::GPU> {
+  cudaStream_t stream;
+};
+
+class ScratchpadAllocator {
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/context.h]: I envision 2 kinds of implementations:
1. Bump allocator with pre-allocated storage
2. LLVM-like allocator - i.e. all allocations are gathered within the buffer object and freed upon allocator's destruction.

File: dali/kernels/tensor_view.h (hunk 7 above)
Comment (mzient) [reply dali/kernels/tensor_view.h]: If we constrain it, then make it `sizeof...(idx)+1 <= dim_`, so we can easily slice. The proposed CalcOffset function already handles that.

File: dali/kernels/tuple_helpers.h (hunk 8 above)
Comment (mzient) [reply dali/kernels/tuple_helpers.h]: Yes, I know. I think this variant would build faster because compiler follows simpler rules in order to realize it. I'd have to check the real cost, though.
BTW, I remember having some problems with declval. Can't recall what that was exactly, but had to do with the fact that declval adds r-value reference.

File: dali/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/CMakeLists.txt]: So, if I understood you correctly, you state, that whoever wants to use Dali, he would need to manually build and install additional version of CMake.

Afaik, Ubuntu 16.04 doesn't get any CMake by default (not to mention docker's `ubuntu16.04:latest`, which doesn't even have `sudo` inside). So user still needs to install cmake.
Ofc, I'm aware, that installing cmake is like `sudo apt-get install cmake`. But! Please look at opencv, if you `sudo apt-get install libopencv-dev` you're installing neither recommended (3.4) nor supported (3.0+) version - you get 2.4.9. Therefore user needs to build it and install manually. Just as with CMake 3.9.

On top of that, if someone really needs CMake for Dali, he want's to build it himself. Therefore, he's not an ordinary python-guy and he knows how to deal with such things...

File: dali/kernels/shape.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/kernels/shape.h]: I agree with the first point. My doubt here is that, when you see `Index`, you rather expect unsigned values. Anywho, isn't using such typedef redundant? We might as well use `int64_t`...

File: dali/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/CMakeLists.txt]: The general idea is not to force the user to make his build env more customized that is really needed. Please remember that OpenCV is recommended but still works with the older version, you can turn off jpeg turbo as well.
What we want to achieve is to make custom DALI build as easy as possible.

File: dali/kernels/shape.h [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+  TensorShape() = default;
+  template <int other_dim>
+  TensorShape(const TensorShape<other_dim> &other) {
+    *this = other;
+  }
+  TensorShape(std::vector<Index> s) : data(std::move(s)) {}
+
+  Index &operator[](int d) { return data[d]; }
+  Index operator[](int d) const { return data[d]; }
+  std::vector<Index> data;  // TODO: replace with something mo
```
Comment (Kh4L) [inline dali/kernels/shape.h]: Something mo?

File: dali/kernels/scratch.h [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  void *Alloc(StorageBackend backend, size_t bytes, size_t alignment) override {
+    auto &A = allocs_[(int)backend];
+    size_t ptr = (size_t)A.next();
+    // Calculate padding needed to satisfy alignmnent requirements
+    size_t padding = (alignment-1) & -ptr;
+    A.New(padding);
+    return A.New(bytes);
+  }
+ private:
+  std::array<BumpAllocator<char>, (size_t)StorageBackend::Count> allocs_;
```
Comment (Kh4L) [inline dali/kernels/scratch.h]: Why not:
```suggestion
  std::array<BumpAllocator<char>, static_cast<size_t>(StorageBackend::Count)> allocs_;
```

File: dali/kernels/util.h [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+#include <cstddef>
+#include <utility>
+
+namespace dali {
+
+using std::size_t;
+
+template <typename Target, typename Source>
+void append(Target &target, const Source &source) {  // NOLINT
```
Comment (Kh4L) [inline dali/kernels/util.h]: You can remove ` // NOLINT` here

File: dali/kernels/scratch.h (hunk 11 above)
Comment (mzient) [reply dali/kernels/scratch.h]: If that's really a question "why": it's too long for what it does _here_.
`static_cast` is invaluable in the context of pointer/reference casting, where it ensures that the types are actually related (and can actually change the numeric value of pointers!) - or in case of templates where it will raise alarm if someone supplies a type that cannot be safely cast, loses qualifiers, etc. When you know that you're casting an enumeration to an integral type, `static_cast<>` is just adding 13 characters to the line length.
TL;DR
So, if I'm to satisfy the linter, I'd prefer constructor-style cast: `size_t(StorageBackend::Count)`.

File: dali/kernels/shape.h (hunk 10 above)
Comment (mzient) [reply dali/kernels/shape.h]: ;) One ctrl+z too much is my guess...

File: dali/kernels/shape.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (klecki) [reply dali/kernels/shape.h]: It is redundant, but it also carries some additional meaning for the programmer.
But strong typedefs would be better.

File: dali/kernels/kernel.h [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+struct Kernel {
+  /// @brief Returns kernel output(s) shape(s) and additional memory requirements
+  ///
+  /// GetRequirements receives full input and output tensor lists and any extra arguments that
+  /// are going to be passed to a subsequent call to Run.
+  ///
+  /// @remarks The inputs are provided mainly to inspect their shapes; actually looking at the
+  /// data may degrade performance severely.
+  ///
+  /// @param context - environoment of the kernel;, cuda stream, batch info, etc.
```
Comment (jantonguirao) [inline dali/kernels/kernel.h]: typo: environment

File: dali/kernels/kernel.h [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+  ///                  At the time of call to GetRequirements, its scratch area is undefined.
+  ///
+  /// @param in1 - example input, consisting of a list of 3D tensors with element type Input1
+  /// @param in2 - example input, consisting of a list of $D tensors with element type Input2
```
Comment (jantonguirao) [inline dali/kernels/kernel.h]: $D? :)

File: dali/kernels/kernel.h [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+  /// Run processes the inputs and populates the pre-allocated output. Output shape is expected
+  /// to match that returned by GetRequirements.
+  ///
+  /// @remarks The inputs are provided mainly to inspect their shapes; actually looking at the
+  /// data may degrade performance severely.
+  ///
+  /// @param context - environment; provides scratch memory, cuda stream, batch info, etc.
+  ///                  Scratch area must satisfy requirements returned by GetRequirements.
+  /// @param in1 - example input, consisting of a list of 3D tensors with element type Input1
+  /// @param in2 - example input, consisting of a list of $D tensors with element type Input2
```
Comment (jantonguirao) [inline dali/kernels/kernel.h]:
```suggestion
  /// @param in2 - example input, consisting of a list of 3D tensors with element type Input2
```

File: dali/kernels/shape.h [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+#include "util.h"
+
+namespace dali {
+
+template <typename T>
+constexpr typename std::enable_if<std::is_fundamental<T>::value, size_t>::type ShapeDim(const T &)
+{ return 1; }
+
+template <typename T, size_t N>
+constexpr int ShapeDim(T (&)[N]) { return int(N); }
```
Comment (jantonguirao) [inline dali/kernels/shape.h]: any reason for not returning size_t?

Comment (mzient) [reply dali/kernels/shape.h]: Negative number of dimensions means "dynamic". But it's more related to https://github.com/NVIDIA/DALI/pull/346/files than this PR, actually.

File: dali/kernels/kernel_params.h [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+namespace kernels {
+
+template <StorageBackend backend, typename ElementType, int sample_dim>
+using InList = TensorListView<backend, const ElementType, sample_dim>;
+
+template <StorageBackend backend, typename ElementType, int sample_dim>
+using OutList = TensorListView<backend, ElementType, sample_dim>;
+
+template <typename ElementType, int sample_dim>
+using InListCPU = InList<StorageBackend::CPU, ElementType, sample_dim>;
```
Comment (klecki) [inline dali/kernels/kernel_params.h]:
```suggestion
using InListCPU = InList<StorageBackend::CPU, const ElementType, sample_dim>;
```

File: dali/kernels/kernel_params.h [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+template <typename ElementType, int sample_dim>
+using InListGPU = InList<StorageBackend::GPU, ElementType, sample_dim>;
```
Comment (klecki) [inline dali/kernels/kernel_params.h]:
```suggestion
using InListGPU = InList<StorageBackend::GPU, const ElementType, sample_dim>;
```

File: dali/kernels/kernel_params.h [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+using OutListGPU = OutList<StorageBackend::GPU, ElementType, sample_dim>;
+
+template <StorageBackend backend, typename ElementType, int sample_dim>
+using InTensor = TensorView<backend, const ElementType, sample_dim>;
+
+template <StorageBackend backend, typename ElementType, int sample_dim>
+using OutTensor = TensorView<backend, ElementType, sample_dim>;
+
+template <typename ElementType, int sample_dim>
+using InTensorCPU = InTensor<StorageBackend::CPU, ElementType, sample_dim>;
```
Comment (klecki) [inline dali/kernels/kernel_params.h]:
```suggestion
using InTensorCPU = InTensor<StorageBackend::CPU, const ElementType, sample_dim>;
```

File: dali/kernels/kernel_params.h [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+
+template <typename ElementType, int sample_dim>
+using InTensorGPU = InTensor<StorageBackend::GPU, ElementType, sample_dim>;
```
Comment (klecki) [inline dali/kernels/kernel_params.h]:
```suggestion
using InTensorGPU = InTensor<StorageBackend::GPU, const ElementType, sample_dim>;
```

File: dali/kernels/tuple_helpers.h [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+
+
+template <int... indices>
+struct seq {};
+
+template <typename seq, typename... T>
+struct pack_indices_helper;
+
+template <int... indices, typename T0, typename... T>
+struct pack_indices_helper<seq<indices...>, T0, T...>
```
Comment (klecki) [inline dali/kernels/tuple_helpers.h]: Why don't you use sizeof...(T) to calculate the size of pack, and than you can generate sequences up to given number?

Also you can do it in log(N), using something like:
```
template <int... A, int... B>
build<seq<A...>, seq<B...>> : seq<A..., (sizeof...(A)+B)...> {}
```

File: dali/kernels/kernel.h [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+    const InListGPU<Input1, 3> &in1,
+    const InListGPU<Input2, 4> &in2,
+    const InTensorCPU<float, 3> &aux);
+
+  /// @brief Runs the kernel
+  ///
+  /// Run processes the inputs and populates the pre-allocated output. Output shape is expected
+  /// to match that returned by GetRequirements.
+  ///
+  /// @remarks The inputs are provided mainly to inspect their shapes; actually looking at the
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/kernel.h]: Copy-paste ;)

File: dali/kernels/kernel.h [hunk 23]
```diff
@@ -0,0 +1,191 @@
...
+
+template <typename Kernel>
+using args = kernel_args<Kernel>;
+
+using Context = KernelContext;
+using Requirements = KernelRequirements;
+
+/// @brief Gets requirements for given Kernel
+/// @param context            - execution environment (without scratch memory)
+/// @param for_type_inference - a placeholder for kernel type, may be a
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/kernel.h]: If not using the placeholder, we'd have to resort to `decltype` :( We might have both.

[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/kernel.h]: +1 for parameter order - I'll swap

File: dali/kernels/kernel_params.h (hunk 17 above)
Comment (mzient) [reply dali/kernels/kernel_params.h]: InList already takes care of that.

File: dali/kernels/kernel_params.h (hunk 18 above)
Comment (mzient) [reply dali/kernels/kernel_params.h]: InList already takes care of that.

File: dali/kernels/kernel_req.h [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+#include <vector>
+#include <algorithm>
+
+namespace dali {
+namespace kernels {
+
+/// @brief Represents requirements for kernel to do its job for given inputs and arguments.
+struct KernelRequirements {
+  // TODO(michalz, klecki) TensorListShape, when we have it
+  std::vector<TensorShape<-1>> output_shapes;
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/kernel_req.h]: That would make specifying requirements for multiple input sets an impossibility. Besides, one would find it quite hard to know when to stop this templatization. Fixed number of outputs? Fixed dimensions? If we try to do the latter, it's no longer an array, but a tuple...

File: dali/kernels/kernel_req.h [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+      if (reuse_scratch)
+        scratch_sizes[i] = std::max(scratch_sizes[i], r.scratch_sizes[i]);
+      else
+        scratch_sizes[i] += r.scratch_sizes[i];
+    }
+    return *this;
+  }
+};
+
+/// @brief A utility class for adding scractchpad requirements with proper alignment,
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/kernel_req.h]: Fixed.

File: dali/kernels/tuple_helpers.h (hunk 21 above)
Comment (mzient) [reply dali/kernels/tuple_helpers.h]: I'll try...

File: dali/kernels/kernel_traits.h [hunk 26]
```diff
@@ -0,0 +1,243 @@
...
+    std::declval<kernel_outputs<Kernel>>(),
+    std::declval<kernel_inputs<Kernel>>(),
+    std::declval<kernel_args<Kernel>>() ))>
+  IsKernelRunnable(Kernel*);
+
+std::false_type IsKernelRunnable(...);
+
+template <typename Kernel>
+std::is_same<KernelRequirements, decltype(apply_all(Kernel::GetRequirements,
+    *static_cast<KernelContext*>(nullptr),  // NOLINT
```
Comment (szalpal) [inline dali/kernels/kernel_traits.h]: Is there a difference between this cast here, and the cast few lines above (the C-style one)?

File: dali/kernels/kernel.h [hunk 27]
```diff
@@ -0,0 +1,191 @@
...
+/// through instantiating check_kernel<KernelType>. If the type does not meet requirements,
+/// static_asserts should produce meaningful diagnostics that will help to rectify the problem.
+template <typename Input1, typename Input2, typename OutputType>
+struct Kernel {
+  /// @brief Returns kernel output(s) shape(s) and additional memory requirements
+  ///
+  /// GetRequirements receives full input and output tensor lists and any extra arguments that
+  /// are going to be passed to a subsequent call to Run.
+  ///
+  /// @remarks The inputs are provided mainly to inspect their shapes; actually looking at the
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/kernel.h]: Nah, we _might_ look at some inputs in some cases. It's just more flexible, although it allows the developer to write suboptimal code. But hey, kernel developers should know better, shouldn't they?

File: dali/kernels/alloc_type.h [hunk 28]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
```
Comment (jantonguirao) [inline dali/kernels/alloc_type.h]: minor comment: Those new files could have 2018 only in the copyright notice

File: dali/kernels/alloc_type.h [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_ALLOC_TYPE_H_
+#define DALI_KERNELS_ALLOC_TYPE_H_
+
+#include <dali/kernels/backend_tags.h>
```
Comment (jantonguirao) [inline dali/kernels/alloc_type.h]: don't we use "" when including internal headers?

File: dali/kernels/context.h [hunk 30]
```diff
@@ -0,0 +1,72 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_CONTEXT_H_
+#define DALI_KERNELS_CONTEXT_H_
+
+#include <cuda_runtime_api.h>
+#include <dali/kernels/tensor_view.h>
```
Comment (jantonguirao) [inline dali/kernels/context.h]: same as above (including with "")

File: dali/kernels/kernel.h [hunk 31]
```diff
@@ -0,0 +1,178 @@
...
+Requirements GetRequirements(
+      Context &context,
+      const std::vector<inputs<Kernel>> &input_sets,
+      const args<Kernel> &args) {
+  check_kernel<Kernel>();
+  if (input_sets.empty())
+    return {};
+
+  Requirements req = GetRequirements<Kernel>(context, input_sets[0], args);
+  for (size_t i = 1; i < input_sets; i++) {
```
Comment (jantonguirao) [inline dali/kernels/kernel.h]: probably you are missing .size() there, but you can even go further with
```suggestion
  for (auto& input : input_sets) {
```

File: dali/kernels/kernel.h [hunk 32]
```diff
@@ -0,0 +1,178 @@
...
+/// @brief Default implementation of execution for multiple input sets
+///
+/// @TODO(michalz) remove references from inputs/outputs/args?
+template <typename Kernel>
+void Run(Context &context,
+      const std::vector<outputs<Kernel> > &output_sets,
+      const std::vector<inputs<Kernel> > &input_sets,
+      const args<Kernel> &args) {
+  check_kernel<Kernel>();
+  assert(output_sets.size() == input_sets.size());
```
Comment (jantonguirao) [inline dali/kernels/kernel.h]: DALI_ENFORCE?

File: dali/kernels/kernel.h [hunk 33]
```diff
@@ -0,0 +1,178 @@
...
+  for (size_t i = 0; i < input_sets.size(); i++) {
+    Run<Kernel>(context, output_sets, input_sets, args);
```
Comment (jantonguirao) [inline dali/kernels/kernel.h]: Shouldn't it be something like 
```suggestion
    Run<Kernel>(context, output_sets[i], input_sets[i], args);
```

File: dali/kernels/kernel.h [hunk 34]
```diff
@@ -0,0 +1,178 @@
...
+  }
+}
+
+}  // namespace kernel
```
Comment (jantonguirao) [inline dali/kernels/kernel.h]: dali::kernels::kernel ?

File: dali/kernels/kernel_req.h [hunk 35]
```diff
@@ -0,0 +1,83 @@
...
+#include <array>
+#include <vector>
+#include <algorithm>
+
+namespace dali {
+namespace kernels {
+
+/// @brief Represents requirements for kernel to do its job for given inputs and arguments.
+struct KernelRequirements {
+  std::vector<TensorListShape<-1>> output_shapes;
```
Comment (jantonguirao) [inline dali/kernels/kernel_req.h]: I'd add a named constant for the meaning of -1

File: dali/kernels/kernel.h (hunk 31 above)
Comment (mzient) [reply dali/kernels/kernel.h]: Yup, will fix...
